dash/dash.py

```diff
@@ -421,18 +422,43 @@ def layout(self):
         return self._layout
 
     def _layout_value(self):
-        if isinstance(self._layout, patch_collections_abc("Callable")):
-            self._cached_layout = self._layout()
-        else:
-            self._cached_layout = self._layout
-        return self._cached_layout
+        return self._layout() if self._layout_is_function else self._layout
 
     @layout.setter
     def layout(self, value):
         _validate.validate_layout_type(value)
-        self._cached_layout = None
+        self._layout_is_function = isinstance(value, patch_collections_abc("Callable"))
         self._layout = value
 
+        # for using flask.has_request_context() to deliver a full layout for
+        # validation inside a layout function - track if a user might be doing this.
+        if self._layout_is_function and not self.validation_layout:
```

Comment:
I wonder if this is something we should turn off if `debug=False` too, or introduce a new `dev_tools_validate_layout` function. Just to reduce the performance hit if there are many small components (e.g. a `dash_html_components.Table`)

Reply:
Good point. Those mostly wouldn't have IDs, so wouldn't be sent to the front end anyway, but with debug off we don't need it at all. In fact I guess we should be able to short-circuit all the front-end validation in that case too.

Reply:
I added clauses to skip this when `suppress_callback_validation` is used, but didn't remove it entirely when `debug=False`. There's more to think about here - on the one hand, without this validation the app behavior can change, and on the other there's a lot more validation we could be skipping if that's what we want to do. I've created #1206 to follow up.